tfrecorder/schema.py

```diff
@@ -0,0 +1,176 @@
+    allowed_values=None)
+
+integer_label = _supported_type(
+    type_name='integer_label',
+    feature_spec=tf.io.FixedLenFeature([], tf.int64),
+    allowed_values=None)
+
+string_label = _supported_type(
+    type_name='string_label',
+    feature_spec=tf.io.FixedLenFeature([], tf.string),
+    allowed_values=None)
+
+image_support_type = _supported_type(
+    type_name='image_support_type',
+    feature_spec=tf.io.FixedLenFeature([], tf.string),
+    allowed_values=None)
+
+# Default schema supports the legacy image_csv format.
+image_csv_schema = frozendict.FrozenOrderedDict({
+    'split': split_key,
+    'image_uri': image_uri,
+    'label': string_label})
+
+
+def get_raw_schema_map(
+    schema_map: Dict[str, collections.namedtuple]
+    ) -> Dict[str, collections.namedtuple]:
+  """Converts a schema to a raw (pre TFT / post image extraction) schema."""
+  raw_schema = {}
+  for k, v in schema_map.items():
+    if v.type_name == "image_uri":
+      raw_schema['image_name'] = image_support_type
+      raw_schema['image'] = image_support_type
+      raw_schema['image_height'] = image_support_type
+      raw_schema['image_width'] = image_support_type
+      raw_schema['image_channels'] = image_support_type
+    else:
+      raw_schema[k] = schema_map[k]
+  return raw_schema
+
+
+def get_tft_coder(columns: List[str],
+                  schema_map: Dict[str, collections.namedtuple]
+                  ) -> tft.coders.CsvCoder:
+  """Gets a TFT CSV Coder.
+
+  Args:
+    columns: Ordered dataframe column names, from df.column.
+    schema_map: Schema map used to infer the schema.
+
+  Returns:
+    tft.coders.CsvCoder
+  """
+  feature_spec = dict()
+
+  # Because the DF column name order may not match the feature_spec order
+  # This maps existing column names to their feature spec (req part of
+  # namedtuple)
+  for col in columns:
+    feature_spec[col] = schema_map[col].feature_spec
+
+
+  metadata = dataset_metadata.DatasetMetadata(
+      schema_utils.schema_from_feature_spec(feature_spec))
+  return tft.coders.CsvCoder(columns,
+                             metadata.schema)
+
+
+def get_key(
+    type_name: str,
+    schema_map: Dict[str, collections.namedtuple]
+    ) -> Union[str, None]:
+  """Gets key of type 'type_name' from schema map if present,
+  otherwise returns None
+
+  Used by the beam pipeline to identify if images or split are present.
+  """
+  for k, v in schema_map.items():
+    if v.type_name == type_name:
+      return k
+  return None
+
+
+def get_raw_feature_spec(columns: List[str],
+                         schema_map: Dict[str, collections.namedtuple]
+                         ) -> Dict[str, tf.io.FixedLenFeature]:
+  """Get's RAW (pre TFT) feature spec."""
+
+  feature_spec = dict()
+
+  # Because the DF column name order may not match the feature_spec order
+  # This maps existing column names to their feature spec (req part of
+  # namedtuple)
+  for col in columns:
```

Comment:
Is this loop necessary if "image_uri" is not specified?

Reply:
I think it is because the schema map and the dataframe columns could be out of order.